internal/tofu/transform_provider.go

```diff
@@ -293,6 +293,25 @@ type ProviderFunctionTransformer struct {
 	ProviderFunctionTracker ProviderFunctionMapping
 }
 
+func createStubProvider(g *Graph, providerVerts map[string]GraphNodeProvider, pAddr addrs.AbsProviderConfig) *NodeUnconfiguredProvider {
+	stubAddr := addrs.AbsProviderConfig{
+		Module:   addrs.RootModule,
+		Provider: pAddr.Provider,
+		Alias:    "",
+	}
+
+	log.Printf("[TRACE] ProviderFunctionTransformer: creating init-only node for %s", stubAddr)
+	stubProvider := &NodeUnconfiguredProvider{
+		&NodeAbstractProvider{
+			Addr: stubAddr,
+		},
+	}
+
+	providerVerts[stubAddr.String()] = stubProvider
+	g.Add(stubProvider)
+	return stubProvider
+}
+
 func (t *ProviderFunctionTransformer) Transform(_ context.Context, g *Graph) error {
```

Comment:
i think we should use this as an opportunity to introduce some good unit tests here.